File: sdk/typescript/src/http.ts
import {
  MorphDBApiError,
  MorphDBNotFoundError,
  MorphDBValidationError,
  MorphDBAuthenticationError,
  MorphDBAuthorizationError,
  MorphDBConflictError,
} from './errors.js';
import type { MorphDBClientOptions } from './types.js';

export class HttpClient {
  private readonly baseUrl: string;
  private readonly options: Required<Pick<MorphDBClientOptions, 'timeout' | 'retryCount' | 'retryDelay'>> &
    MorphDBClientOptions;

  constructor(baseUrl: string, options: MorphDBClientOptions = {}) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.options = {
      timeout: options.timeout ?? 30000,
      retryCount: options.retryCount ?? 3,
      retryDelay: options.retryDelay ?? 1000,
      ...options,
    };
  }

  private getHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      Accept: 'application/json',
    };

    if (this.options.projectId) {
      headers['X-Project-Id'] = this.options.projectId;
    }

    if (this.options.apiKey) {
      headers['X-API-Key'] = this.options.apiKey;
    }

    if (this.options.jwtToken) {
      headers['Authorization'] = `Bearer ${this.options.jwtToken}`;
    }

    return headers;
  }

  async request<T>(
    method: string,
    path: string,
    body?: unknown,
    options?: { headers?: Record<string, string>; returnBlob?: boolean }
  ): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const headers = { ...this.getHeaders(), ...options?.headers };

    let lastError: Error | undefined;

    for (let attempt = 0; attempt <= this.options.retryCount; attempt++) {
      try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.options.timeout);

        const response = await fetch(url, {
          method,
          headers,
          body: body ? JSON.stringify(body) : undefined,
          signal: controller.signal,
        });

        clearTimeout(timeoutId);

        if (!response.ok) {
          const responseBody = await response.text();
          throw this.createError(response.status, responseBody, url);
        }

        if (options?.returnBlob) {
          return (await response.blob()) as T;
        }

        const text = await response.text();
        if (!text) {
          return undefined as T;
        }

        return JSON.parse(text) as T;
      } catch (error) {
        lastError = error as Error;

        // Don't retry on client errors
        if (error instanceof MorphDBApiError && error.statusCode >= 400 && error.statusCode < 500) {
          throw error;
        }

        // Wait before retry
        if (attempt < this.options.retryCount) {
          await this.delay(this.options.retryDelay * Math.pow(2, attempt));
        }
      }
    }

    throw lastError;
  }

  async get<T>(path: string): Promise<T> {
    return this.request<T>('GET', path);
  }

  async post<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>('POST', path, body);
  }

  async put<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>('PUT', path, body);
  }

  async patch<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>('PATCH', path, body);
  }

  async delete<T>(path: string): Promise<T> {
    return this.request<T>('DELETE', path);
  }

  async postFormData<T>(path: string, formData: FormData): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const headers = { ...this.getHeaders() };
    delete headers['Content-Type']; // Let browser set it for FormData

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.options.timeout);

    const response = await fetch(url, {
      method: 'POST',
      headers,
      body: formData,
      signal: controller.signal,
    });

    clearTimeout(timeoutId);

    if (!response.ok) {
      const responseBody = await response.text();
      throw this.createError(response.status, responseBody, url);
    }

    return response.json() as Promise<T>;
  }

  async getBlob(path: string): Promise<Blob> {
    return this.request<Blob>('GET', path, undefined, { returnBlob: true });
  }

  private createError(status: number, responseBody: string, url: string): MorphDBApiError {
    switch (status) {
      case 404:
        return new MorphDBNotFoundError(`Resource not found: ${url}`, responseBody);
      case 400:
        return new MorphDBValidationError('Validation failed', undefined, responseBody);
      case 401:
        return new MorphDBAuthenticationError('Authentication required', responseBody);
      case 403:
        return new MorphDBAuthorizationError('Access denied', responseBody);
      case 409:
        return new MorphDBConflictError('Resource conflict', responseBody);
      default:
        return new MorphDBApiError(`Request failed with status ${status}`, status, undefined, responseBody);
    }
  }

  private delay(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  setProjectId(projectId: string): void {
    this.options.projectId = projectId;
  }

  setApiKey(apiKey: string): void {
    this.options.apiKey = apiKey;
  }

  setJwtToken(jwtToken: string): void {
    this.options.jwtToken = jwtToken;
  }
}

File: sdk/python/morphdb/http.py
"""MorphDB SDK HTTP client utilities."""

import asyncio
from typing import Any

import httpx

from morphdb.exceptions import (
    MorphDBApiError,
    MorphDBAuthenticationError,
    MorphDBAuthorizationError,
    MorphDBConflictError,
    MorphDBConnectionError,
    MorphDBNotFoundError,
    MorphDBValidationError,
)


class HttpClient:
    """HTTP client with retry logic and error handling."""

    def __init__(
        self,
        base_url: str,
        project_id: str,
        api_key: str | None = None,
        jwt_token: str | None = None,
        timeout: float = 30.0,
        retry_count: int = 3,
        retry_delay: float = 1.0,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.project_id = project_id
        self.api_key = api_key
        self.jwt_token = jwt_token
        self.timeout = timeout
        self.retry_count = retry_count
        self.retry_delay = retry_delay
        self._client: httpx.AsyncClient | None = None

    def _get_headers(self) -> dict[str, str]:
        """Get request headers."""
        headers = {
            "Content-Type": "application/json",
            "X-Project-Id": self.project_id,
        }
        if self.api_key:
            headers["X-API-Key"] = self.api_key
        if self.jwt_token:
            headers["Authorization"] = f"Bearer {self.jwt_token}"
        return headers

    async def _get_client(self) -> httpx.AsyncClient:
        """Get or create the HTTP client."""
        if self._client is None or self._client.is_closed:
            self._client = httpx.AsyncClient(
                base_url=self.base_url,
                headers=self._get_headers(),
                timeout=self.timeout,
            )
        return self._client

    async def close(self) -> None:
        """Close the HTTP client."""
        if self._client is not None and not self._client.is_closed:
            await self._client.aclose()
            self._client = None

    def set_project_id(self, project_id: str) -> None:
        """Update project ID."""
        self.project_id = project_id

    def set_api_key(self, api_key: str) -> None:
        """Update API key."""
        self.api_key = api_key

    def set_jwt_token(self, jwt_token: str) -> None:
        """Update JWT token."""
        self.jwt_token = jwt_token

    async def _handle_response(self, response: httpx.Response) -> Any:
        """Handle HTTP response and raise appropriate exceptions."""
        if response.status_code >= 200 and response.status_code < 300:
            if response.status_code == 204:
                return None
            try:
                return response.json()
            except Exception:
                return response.text

        response_body = response.text
        error_message = f"HTTP {response.status_code}"

        try:
            error_data = response.json()
            if isinstance(error_data, dict):
                error_message = error_data.get("message", error_message)
                error_code = error_data.get("code")
                errors = error_data.get("errors")
        except Exception:
            error_code = None
            errors = None

        if response.status_code == 400:
            raise MorphDBValidationError(error_message, errors, response_body)
        elif response.status_code == 401:
            raise MorphDBAuthenticationError(error_message, response_body)
        elif response.status_code == 403:
            raise MorphDBAuthorizationError(error_message, response_body)
        elif response.status_code == 404:
            raise MorphDBNotFoundError(error_message, response_body)
        elif response.status_code == 409:
            raise MorphDBConflictError(error_message, response_body)
        else:
            raise MorphDBApiError(error_message, response.status_code, error_code, response_body)

    async def _request_with_retry(
        self,
        method: str,
        path: str,
        **kwargs: Any,
    ) -> Any:
        """Execute request with retry logic."""
        client = await self._get_client()
        last_exception: Exception | None = None

        for attempt in range(self.retry_count):
            try:
                # Update headers for each request in case credentials changed
                headers = kwargs.pop("headers", {})
                headers.update(self._get_headers())
                kwargs["headers"] = headers

                response = await client.request(method, path, **kwargs)
                return await self._handle_response(response)
            except (httpx.ConnectError, httpx.TimeoutException) as e:
                last_exception = MorphDBConnectionError(str(e))
                if attempt < self.retry_count - 1:
                    await asyncio.sleep(self.retry_delay * (attempt + 1))
            except MorphDBApiError:
                raise
            except Exception as e:
                raise MorphDBConnectionError(str(e)) from e

        if last_exception:
            raise last_exception
        raise MorphDBConnectionError("Request failed after retries")

    async def get(self, path: str, params: dict[str, Any] | None = None) -> Any:
        """Execute GET request."""
        return await self._request_with_retry("GET", path, params=params)

    async def post(
        self,
        path: str,
        data: dict[str, Any] | None = None,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Execute POST request."""
        return await self._request_with_retry("POST", path, json=data, params=params)

    async def put(
        self,
        path: str,
        data: dict[str, Any] | None = None,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Execute PUT request."""
        return await self._request_with_retry("PUT", path, json=data, params=params)

    async def patch(
        self,
        path: str,
        data: dict[str, Any] | None = None,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Execute PATCH request."""
        return await self._request_with_retry("PATCH", path, json=data, params=params)

    async def delete(self, path: str, params: dict[str, Any] | None = None) -> Any:
        """Execute DELETE request."""
        return await self._request_with_retry("DELETE", path, params=params)

    async def post_file(
        self,
        path: str,
        file_content: bytes,
        filename: str,
        content_type: str = "application/octet-stream",
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Execute POST request with file upload."""
        files = {"file": (filename, file_content, content_type)}
        client = await self._get_client()

        headers = self._get_headers()
        # Remove Content-Type for multipart form data
        headers.pop("Content-Type", None)

        response = await client.post(path, files=files, params=params, headers=headers)
        return await self._handle_response(response)

    async def get_bytes(self, path: str, params: dict[str, Any] | None = None) -> bytes:
        """Execute GET request and return raw bytes."""
        client = await self._get_client()
        response = await client.get(path, params=params, headers=self._get_headers())

        if response.status_code >= 200 and response.status_code < 300:
            return response.content

        await self._handle_response(response)
        return b""  # Never reached, but satisfies type checker
